### PR TITLE
feat: Report application errors with details

### PR DESCRIPTION
**Pull request by erezrokah**
#### Summary

Related to https://github.com/cloudquery/plugin-sdk/issues/462.
Draft PR as meant to serve as an example on how to improve error reporting and handling.
I'm using `errdetails` but we could also write our own protobuf messages instead (don't think we should).

---

Use the following steps to ensure your PR is ready to be reviewed

- [ ] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [ ] Run `go fmt` to format your code 🖊
- [ ] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [ ] Update or add tests 🧪
- [ ] Ensure the status checks below are successful ✅

### CODE REVIEW

File: clients/destination.go [hunk 1]
```diff
@@ -300,7 +301,20 @@ func (c *DestinationClient) Write2(ctx context.Context, tables schema.Tables, so
 	}
 	_, err = saveClient.CloseAndRecv()
 	if err != nil {
-		return fmt.Errorf("failed to CloseAndRecv client: %w", err)
+		st := status.Convert(err)
+		errorWithDetails := err.Error()
```
Comment (erezrokah) [inline clients/destination.go:305]: I'm concatenating errors into one string, but we can decide to do something else

File: internal/servers/destinations.go [hunk 2]
```diff
@@ -86,36 +94,42 @@ func (s *DestinationServer) Write2(msg pb.Destination_Write2Server) error {
...
 		}
 		var resource schema.DestinationResource
-		if err := json.Unmarshal(r.Resource, &resource); err != nil {
+		if unmarshalError := json.Unmarshal(r.Resource, &resource); unmarshalError != nil {
 			close(resources)
-			if err := eg.Wait(); err != nil {
-				s.Logger.Error().Err(err).Msg("failed to unmarshal resource. failed to wait for plugin")
+			errWithDetails := addErrorDetails(errors.New("failed to unmarshal resource"), unmarshalError.Error())
+			if writeError := eg.Wait(); writeError != nil {
+				errWithDetails = addErrorDetails(errors.New("plugin write failed and failed to unmarshal resource"), writeError.Error())
```
Comment (erezrokah) [inline internal/servers/destinations.go:116]: `Details` is an array so we can send multiple errors

File: internal/servers/destinations.go [hunk 3]
```diff
@@ -86,36 +94,42 @@ func (s *DestinationServer) Write2(msg pb.Destination_Write2Server) error {
...
 			if err := eg.Wait(); err != nil {
-				s.Logger.Error().Err(err).Msg("failed to wait")
+				return addErrorDetails(errors.New("plugin write failed"), err.Error())
+			}
+			ctxError := ctx.Err()
+			if ctxError == nil {
+				return nil
 			}
-			return ctx.Err()
+			return addErrorDetails(errors.New("context error"), ctxError.Error())
```
Comment (erezrokah) [inline internal/servers/destinations.go:132]: This is only an example, we should be using https://github.com/cloudquery/plugin-sdk/pull/486 instead

File: clients/destination.go [hunk 4]
```diff
@@ -300,7 +301,20 @@ func (c *DestinationClient) Write2(ctx context.Context, tables schema.Tables, so
 	}
 	_, err = saveClient.CloseAndRecv()
 	if err != nil {
-		return fmt.Errorf("failed to CloseAndRecv client: %w", err)
+		st := status.Convert(err)
```
Comment (erezrokah) [inline clients/destination.go:304]: We should probably put this into a function and replace https://github.com/cloudquery/plugin-sdk/pull/477 so we can get both the message and the details

File: clients/destination.go [hunk 5]
```diff
@@ -300,7 +301,20 @@ func (c *DestinationClient) Write2(ctx context.Context, tables schema.Tables, so
...
+		errorWithDetails := st.Message()
+		if len(st.Details()) > 0 {
+			errorWithDetails += "\nDetails:\n"
```
Comment (candiduslynx) [inline clients/destination.go:307]: can we just utilize [prototext](https://pkg.go.dev/google.golang.org/protobuf/encoding/prototext)?

Comment (erezrokah) [reply clients/destination.go:307]: We can try good catch. Please note this is only one example on how to report the errors. We might want to do something else than strings concatenation

**Comment by erezrokah** [comment]
Closing this POC for now

**Pull request closed without merging**